src/devices/src/virtio/block/request.rs

```diff
@@ -0,0 +1,362 @@
+            .open(PathBuf::from(&disk_image_path))?;
+        let disk_size = disk_image.seek(SeekFrom::End(0))? as u64;
+
+        // We only support disk size, which uses the first two words of the configuration space.
+        // If the image is not a multiple of the sector size, the tail bits are not exposed.
+        if disk_size % SECTOR_SIZE != 0 {
+            // warn!(
+            //     "Disk size {} is not a multiple of sector size {}; \
+            //       the remainder will not be visible to the guest.",
+            //     disk_size, SECTOR_SIZE
+            // );
+        }
+
+        Ok(Self {
+            nsectors: disk_size >> SECTOR_SHIFT,
+            image_id: Self::build_disk_image_id(&disk_image),
+            file_path: disk_image_path,
+            file: disk_image,
+        })
+    }
+
+    pub fn file_mut(&mut self) -> &mut File {
+        &mut self.file
+    }
+
+    pub fn nsectors(&self) -> u64 {
+        self.nsectors
+    }
+
+    pub fn image_id(&self) -> &[u8] {
+        &self.image_id
+    }
+
+    fn build_device_id(disk_file: &File) -> result::Result<String, Error> {
+        let blk_metadata = disk_file.metadata().map_err(Error::GetFileMetadata)?;
+        // This is how kvmtool does it.
+        let device_id = format!(
+            "{}{}{}",
+            blk_metadata.st_dev(),
+            blk_metadata.st_rdev(),
+            blk_metadata.st_ino()
+        );
+        Ok(device_id)
+    }
+
+    fn build_disk_image_id(disk_file: &File) -> Vec<u8> {
+        let mut default_id = vec![0; VIRTIO_BLK_ID_BYTES as usize];
+        match Self::build_device_id(disk_file) {
+            Err(_) => {
+                // warn!("Could not generate device id. We'll use a default.");
+            }
+            Ok(m) => {
+                // The kernel only knows to read a maximum of VIRTIO_BLK_ID_BYTES.
+                // This will also zero out any leftover bytes.
+                let disk_id = m.as_bytes();
+                let bytes_to_copy = cmp::min(disk_id.len(), VIRTIO_BLK_ID_BYTES as usize);
+                default_id[..bytes_to_copy].clone_from_slice(&disk_id[..bytes_to_copy])
+            }
+        }
+        default_id
+    }
+
+    /// Backing file path.
+    pub fn file_path(&self) -> &String {
+        &self.file_path
+    }
+
+    /// Provides vec containing the virtio block configuration space
+    /// buffer. The config space is populated with the disk size based
+    /// on the backing file size.
+    pub fn virtio_block_config_space(&self) -> Vec<u8> {
+        // The config space is little endian.
+        let mut config = Vec::with_capacity(CONFIG_SPACE_SIZE);
+        for i in 0..CONFIG_SPACE_SIZE {
+            config.push((self.nsectors >> (8 * i)) as u8);
+        }
+        config
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RequestType {
+    In,
+    Out,
+    Flush,
+    GetDeviceID,
+    Unsupported(u32),
+}
+
+impl From<u32> for RequestType {
```

Comment:
I think it'd make more sense to couple this RFC with the other block device-related RFC https://github.com/rust-vmm/vm-virtio/pull/29 and use the abstractions from there.

Reply:
It could be a good moment to do this, indeed. Do we plan to do this in the same PR or another one?

Reply:
I've done this (by pulling the changes from your PR in the `vm-virtio` branch used as a depenency right now) in #41.